### PR TITLE
adding bslstl::bad_optional_access exception

### PR DESCRIPTION
**Pull request by NinaRanns**
Implementation of bslstl::bad_optional_access exception needed for implementation of bsl::optional::value()

### CODE REVIEW

File: groups/bsl/bslstl/bslstl_badoptionalaccess.h [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+#ifdef BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY
+#else
+namespace BloombergLP {
+namespace bslstl {
+
+                       // =======================
+                       // class BadOptionalAccess
+                       // =======================
+
+class BadOptionalAccess : public native_std::exception {
```
Comment (mversche) [inline groups/bsl/bslstl/bslstl_badoptionalaccess.h]: I believe the use of native_std is now an anarchronism (as is the corresponding #include)

File: groups/bsl/bslstl/bslstl_badoptionalaccess.h [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+#ifndef INCLUDED_BSLSTL_BADOPTIONALACCESS
+#define INCLUDED_BSLSTL_BADOPTIONALACCESS
+
+#include <bsls_ident.h>
+BSLS_IDENT("$Id: $")
+
+//@PURPOSE: Provide an exception class thrown by 'bsl::optional'.
+//
+//@CLASSES:
+//  bslstl::BadOptionalAccess: exception possibly thrown by 'bsl::optional'
```
Comment (mversche) [inline groups/bsl/bslstl/bslstl_badoptionalaccess.h]: I'm not sure this set of classes makes sense at the moment:  `bslstl::BadOptionalAccess` is only available on older language dialects, so there seems to be little purpose to making that type publicly accessible.  And implementing `bslstl::BadOptionalAccess` for older language dialects would be confusing, since the relationship with `bsl::bad_optional_access` would depend on the language dialect.

So I think there is little to be gained, and likely confusion created, by making `bslstl::BadOptionalAccess` public.

We could have a component-private type, `bslstl::BadOptionalAccess_Imp` (which would not appear in the list of classes), which would be aliased to `bsl::bad_optional_access`.

At that point though, I would simply define `bsl::bad_optional_access` in-place, and forgo the imp type

**Review by mversche — CHANGES_REQUESTED** [review]
A couple quick review comments

File: groups/bsl/bslstl/bslstl_badoptionalaccess.h (hunk 1 above)
Comment (NinaRanns) [reply groups/bsl/bslstl/bslstl_badoptionalaccess.h:64]: do you mean BadOptionalAccess (or in-place bad_optional_access, as suggested below) should inherit directly from std::exception or is there something else I should use ?

File: groups/bsl/bslstl/bslstl_badoptionalaccess.h [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+//  bsl::bad_optional_access: alias to exception type thrown by 'bsl::optional'
```
Comment (khlebnikov) [inline groups/bsl/bslstl/bslstl_badoptionalaccess.h:12]: test comment

File: groups/bsl/bslstl/bslstl_badoptionalaccess.h (hunk 1 above)
Comment (NinaRanns) [reply groups/bsl/bslstl/bslstl_badoptionalaccess.h]: changed to inherit directly from std::exception

File: groups/bsl/bslstl/bslstl_badoptionalaccess.h (hunk 2 above)
Comment (NinaRanns) [reply groups/bsl/bslstl/bslstl_badoptionalaccess.h]: defined bsl::bad_optional_access in-place, and removed bslstl::BadOptionalAccess

File: groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+        //: 3 For concern 3, in step 2, use a const qualified
+        //:   'bad_optional_access' object.
+        //
+        // Testing:
+        //   bad_optional_access();
+        //   const char *what() const;
+        // --------------------------------------------------------------------
+        if (verbose) printf("\nDEFAULT CONSTRUCTION AND 'what' METHOD"
+                            "\n======================================\n");
+#ifndef BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY
```
Comment (mversche) [inline groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp:173]: This test does not seem C++17 specific.

File: groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+        // N4835: 20.6.5 Class 'bad_optional_access' [optional.bad.access]
+        //..
+        //     namespace std {
+        //       class bad_optional_access: public std::exception {
+        //       public:
+        //         const char* what() const noexcept override;
+        //       };
+        //     } // namespace std
+        //..
+#ifndef BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY
```
Comment (mversche) [inline groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp:214]: This test is also not C++17 specific (but is C++11 specific)

File: groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+        // Plan:
+        //: 1 Perform and ad-hoc test of the primary modifiers and accessors.
+        //
+        // Testing:
+        //   BREATHING TEST
+        // --------------------------------------------------------------------
+
+        if (verbose) printf("\nBREATHING TEST"
+                            "\n==============\n");
+#ifndef BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY
```
Comment (mversche) [inline groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp]: This test is not C++17 specific

File: groups/bsl/bslstl/bslstl_badoptionalaccess.h [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        // used here, as a number of standard libraries declare 'what' in the
+        // base 'exception' class explicitly with the no-throw specification,
+        // even in a build that may not recognize exceptions.
+};
+
+// ============================================================================
+//                           INLINE DEFINITIONS
+// ============================================================================
+
+                       // -----------------------
```
Comment (mversche) [inline groups/bsl/bslstl/bslstl_badoptionalaccess.h]: fix banner

File: groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp (hunk 4 above)
Comment (NinaRanns) [reply groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp:173]: In C++17, standard library provides the implementation of this exception type. Because C++17 bsl::optional for non allocator aware value types uses std::optional as the underlying type (which may throw std::bad_optional_access) , C++17 bsl::bad_optional_access is just an alias for std::bad_optional_access. It seems pointless to run the tests for the standard exception type. 
Let me know if you want these tests to run even if bsl::bad_optional_access is an alias for std::bad_optional_access. (note that test 4 checks for type equality in C++17)

Comment (mversche) [reply groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp:173]: I see what you're saying.  That said, it's more complex to not test the type than to test it, and It would be a clear demonstration that our definition of the type matches the standard (because you can see them passing the same set of tests).

File: groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp (hunk 5 above)
Comment (NinaRanns) [reply groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp:214]: Same as above applies for C++17. I'm going to assume the 'but is C++11 specific' comment is related to noexcept specifier being a C++11 feature, however, both BSLS_KEYWORD_NOEXCEPT_AVAILABLE and BSLS_KEYWORD_NOEXCEPT_AVAILABLE macros are defined in such a way that this test works in all C++ versions and there is no need to check for availability of noexcept specifier. Would you still prefer a BSLS_COMPILERFEATURES_SUPPORT_NOEXCEPT guard ?

File: groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp (hunk 6 above)
Comment (NinaRanns) [reply groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp]: same as above

File: groups/bsl/bslstl/bslstl_badoptionalaccess.h (hunk 7 above)
Comment (NinaRanns) [reply groups/bsl/bslstl/bslstl_badoptionalaccess.h]: banner lines extended

File: groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp (hunk 4 above)
Comment (NinaRanns) [reply groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp:173]: Understood. I'll make the tests non C++ version specific.

File: groups/bsl/bslstl/bslstl_badoptionalaccess.h [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+//@PURPOSE: Provide an exception class thrown by 'bsl::optional'.
+//
+//@CLASSES:
+//  bsl::bad_optional_access: exception type thrown by 'bsl::optional'
+//
+//@SEE_ALSO: bslstl_optional, bslstl_stdexceptionutil
+//
+//@DESCRIPTION: This component provides a 'bsl::bad_optional_access' exception
+// class.  This exception is thrown by  'bsl::optional::value' when accessing a
```
Comment (khlebnikov) [inline groups/bsl/bslstl/bslstl_badoptionalaccess.h]: Unnecessary double space before `'bsl::optional::value'`

File: groups/bsl/bslstl/bslstl_badoptionalaccess.h [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+// 'bsl::optional' object that does not contain a value.  If 'std::optional'
+// implementation is available, 'bsl::bad_optional_access' is an alias to
+// 'std::bad_optional_access'.
+//
+
+// Prevent 'bslstl' headers from being included directly in 'BSL_OVERRIDES_STD'
+// mode.  Doing so is unsupported, and is likely to cause compilation errors.
+#if defined(BSL_OVERRIDES_STD) && !defined(BOS_STDHDRS_PROLOGUE_IN_EFFECT)
+#error "include <bsl_memory.h> instead of <bslstl_badoptionalaccess.h> in \
```
Comment (khlebnikov) [inline groups/bsl/bslstl/bslstl_badoptionalaccess.h]: `bsl_memory.h` seems wrong.

File: groups/bsl/bslstl/bslstl_badoptionalaccess.h [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+#include <bsls_exceptionutil.h>
+#include <bsls_keyword.h>
+
+#include <exception>
+
+#ifdef BDE_BUILD_TARGET_EXC
+
+#ifdef BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY
+#include <optional> // 'std::bad_optional_access' if defined
```
Comment (khlebnikov) [inline groups/bsl/bslstl/bslstl_badoptionalaccess.h]: `if defined` seems odd. Can it *not* be defined in `<optional>`?

File: groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+      case 3: {
+        // --------------------------------------------------------------------
+        // DEFAULT CONSTRUCTION AND 'what' METHOD
+        //
+        // Concerns:
+        //: 1 A 'bad_optional_access' object can be default constructed.
+        //: 2 Invoking a 'what' method on a 'bad_optional_access' object
+        //:   returns a "bad_optional_access" string.
+        //: 3 'what' method can be invoked on a const 'bad_optional_access'
+        //:   object
```
Comment (khlebnikov) [inline groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp]: Add `.`

File: groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+        //       public:
+        //         const char* what() const noexcept override;
+        //       };
+        //     } // namespace std
+        //..
+#ifndef BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY
+        ASSERT(BSLS_KEYWORD_NOEXCEPT_AVAILABLE
+            == BSLS_KEYWORD_NOEXCEPT_OPERATOR(bsl::bad_optional_access()));
+
+        bsl::bad_optional_access    b;
```
Comment (khlebnikov) [inline groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp]: Remove unnecessary whitespace before `b`

File: groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+        if (verbose) printf("\nBREATHING TEST"
+                            "\n==============\n");
+#ifndef BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY
+        bsl::bad_optional_access  b;
+        const std::exception     *ptr = &b;
+
+        ASSERT(0 == strcmp("bad_optional_access", b.what()));
+        ASSERT(0 == strcmp("bad_optional_access", ptr->what()));
+#endif
+      }break;
```
Comment (khlebnikov) [inline groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp]: Missing space before `break`

**Review by khlebnikov — COMMENTED** [review]
Minor pedantic comments

File: groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp (hunk 4 above)
Comment (NinaRanns) [reply groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp:173]: Actually, I can't make this one work in C++17 as the string returned from what() is implementation specific. I've added a comment instead.

File: groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp (hunk 5 above)
Comment (NinaRanns) [reply groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp:214]: This one also can't have C++17 guards removed. At least one implementation (gcc) doesn't have the correct noexcept specification. Issue raised with the gcc library team. Comment added instead.

File: groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp (hunk 6 above)
Comment (NinaRanns) [reply groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp]: removed the guards and changed the test to check only that what() returns a non null string.

File: groups/bsl/bslstl/bslstl_badoptionalaccess.h (hunk 8 above)
Comment (NinaRanns) [reply groups/bsl/bslstl/bslstl_badoptionalaccess.h]: fixed

File: groups/bsl/bslstl/bslstl_badoptionalaccess.h (hunk 9 above)
Comment (NinaRanns) [reply groups/bsl/bslstl/bslstl_badoptionalaccess.h]: hmm... this is pretty much a remnant of me using bslstl_badweakptr.h as a source for this file. I don't think this whole #if block makes sense here. Removed

File: groups/bsl/bslstl/bslstl_badoptionalaccess.h (hunk 10 above)
Comment (NinaRanns) [reply groups/bsl/bslstl/bslstl_badoptionalaccess.h]: `<optional>` header itself does not exist before C++17

File: groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp (hunk 11 above)
Comment (NinaRanns) [reply groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp]: fixed

File: groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp (hunk 12 above)
Comment (NinaRanns) [reply groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp]: fixed

File: groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp (hunk 13 above)
Comment (NinaRanns) [reply groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp]: fixed

File: groups/bsl/bslstl/bslstl_badoptionalaccess.h (hunk 10 above)
Comment (mversche) [reply groups/bsl/bslstl/bslstl_badoptionalaccess.h]: I think slava was saying the comment doesn't make sense..?

File: groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp (hunk 5 above)
Comment (mversche) [reply groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp:214]: I see what you're saying - if it compiles and runs on all platforms we're all good.

**Review by mversche — CHANGES_REQUESTED** [review]

File: groups/bsl/bslstl/bslstl_badoptionalaccess.h (hunk 10 above)
Comment (NinaRanns) [reply groups/bsl/bslstl/bslstl_badoptionalaccess.h]: oh, I see. Changed to "implementation of 'std::bad_optional_access' ". Let me know if you want it worded differently.

Comment (khlebnikov) [reply groups/bsl/bslstl/bslstl_badoptionalaccess.h]: We often use just a simple "for" in such comments, so 

```c++
#include <optional>  // for 'std::bad_optional_access'
```

Comment (NinaRanns) [reply groups/bsl/bslstl/bslstl_badoptionalaccess.h]: Noted. Changed as suggested.

**Review by mversche — APPROVED** [review]
